src/jdk.compiler/share/classes/com/sun/tools/javac/util/Iterators.java

```diff
@@ -43,63 +44,49 @@ public static <I, O> Iterator<O> createCompoundIterator(Iterable<I> inputs, Func
         return new CompoundIterator<>(inputs, converter);
     }
 
-    private static class CompoundIterator<I, O> implements Iterator<O> {
+    private static final class CompoundIterator<I, O> implements Iterator<O> {
 
         private final Iterator<I> inputs;
         private final Function<I, Iterator<O>> converter;
-        @SuppressWarnings("unchecked")
-        private Iterator<O> currentIterator = EMPTY;
+        private Iterator<O> currentIterator = Collections.emptyIterator();
 
         public CompoundIterator(Iterable<I> inputs, Function<I, Iterator<O>> converter) {
             this.inputs = inputs.iterator();
             this.converter = converter;
         }
 
+        @Override
         public boolean hasNext() {
             if (currentIterator != null && !currentIterator.hasNext()) {
                 update();
             }
             return currentIterator != null;
         }
 
+        @Override
         public O next() {
-            if (currentIterator == EMPTY && !hasNext()) {
+            if (!hasNext()) {
                 throw new NoSuchElementException();
             }
             return currentIterator.next();
         }
 
-        public void remove() {
-            throw new UnsupportedOperationException();
-        }
-
         private void update() {
             while (inputs.hasNext()) {
                 currentIterator = converter.apply(inputs.next());
-                if (currentIterator.hasNext()) return;
+                if (currentIterator.hasNext()) return; // implicit null check
             }
             currentIterator = null;
         }
     }
 
-    @SuppressWarnings("rawtypes")
-    private static final Iterator EMPTY = new Iterator() {
-        public boolean hasNext() {
-            return false;
-        }
-
-        @Override
-        public Object next() {
-            return null;
-        }
-    };
-
     public static <E> Iterator<E> createFilterIterator(Iterator<E> input, Predicate<E> test) {
-        return new Iterator<E>() {
+        return new Iterator<>() {
             private E current = update();
             private E update () {
                 while (input.hasNext()) {
                     E sym = input.next();
+                    assert sym != null; // otherwise, false positive EOF
```

Comment:
javac coding style is to use the `Assert` class, not `assert` statement

Reply:
If you are considering an unconditional null check here, then you might want to consider a dedicated EOF flag instead. Alternatively, I can remove the `assert`, but leave the comment; for the next reader.

It's probably the case that iteration elements are never `null`. If they were `null`, we would either notice false positive EOF or see NPEs thrown from predicates.

Reply:
You were the one to add the `assert` ;-) 
I just made a statement that overall, the coding style in javac has been to use `Assert` instead of `assert`, at least in those cases where it is appropriate to make an assertion.
Looking at the code in more detail, it's not clear to me that either an assertion or a comment (here) is required. If you want a comment, I'd suggest a more declarative comment on the class itself.

Reply:
Addressed in 140afc8.